packages/cli/src/platform/login.ts

```diff
@@ -1,12 +1,111 @@
+
+interface AuthResult {
+  token: string
+  user: {
+    id: string
+    displayName: string
+    handle: string
+    email: string
+  }
+}
 
 export class Login implements Command {
   public static new(): Login {
     return new Login()
   }
 
-  public async parse(argv: string[]) {
-    await Promise.resolve('todo')
-    return JSON.stringify(argv)
+  public async parse() {
+    console.log('Authenticating to Prisma Platform CLI via browser')
+
+    const server = http.createServer()
+    const { port } = await listen(server, 0, '127.0.0.1')
+
+    const authRedirectUri = `http://localhost:${port}`
+    const authSigninUrl = await generateAuthSigninUrl({ connection: `github`, redirectTo: authRedirectUri })
+
+    console.log('Visit the following URL in your browser to authenticate:')
+    console.log(underline(authSigninUrl.href))
+
+    try {
+      const [authResult] = await Promise.all([
+        new Promise<AuthResult>((resolve, reject) => {
+          server.once('request', (req, res) => {
+            server.close()
+            res.setHeader('connection', 'close')
+            const searchParams = new URL(req.url || '/', 'http://localhost').searchParams
+            const token = searchParams.get('token') ?? ''
+            const error = searchParams.get('error')
+            const location = new URL(`${platformConsoleUrl}/auth/cli`)
+
+            if (error) {
+              location.pathname += '/error'
+              location.searchParams.set('error', error)
+              reject(new Error(error))
+            } else {
+              // TODO: Consider getting the user via Console API instead of passing it via query params
+              const user = parseUser(searchParams.get('user') ?? '')
+              if (user) {
+                location.pathname += '/success'
+                location.searchParams.set('email', user.email)
+                resolve({ token, user })
+              } else {
+                location.pathname += '/error'
+                location.searchParams.set('error', 'Invalid user')
+                reject(new Error('Invalid user'))
+              }
+            }
+
+            res.statusCode = 302
+            res.setHeader('location', location.href)
+            res.end()
+          })
+          server.once('error', reject)
+        }),
+        open(authSigninUrl.href),
+      ])
+
+      await writeAuthConfig({ token: authResult.token })
+      console.log('Authenticated successfully as:')
+      console.log(JSON.stringify(authResult.user, null, 4))
+      return ''
+    } catch (error) {
+      throw new Error(`Authentication failed: ${isError(error) ? error.message : ''}`)
+    }
+  }
+}
+
+const generateAuthSigninUrl = async (params: { connection: string; redirectTo: string }) => {
+  const prismaClientVersion = await getInstalledPrismaClientVersion().catch(() => null)
+  const state = { client: `prisma@${prismaClientVersion}`, ...params }
```

Comment:
do we need the `prisma@` ? would the consumer of this need to remove it later to get the version?

Reply:
I was bouncing between two options:
1.  
```ts
const state = { client: `prisma@<version number>`}
```
2.
```ts
const state = { client: `prisma`, version: <version number>}
```
@jasonkuhrt already established a pattern here and I thought to follow the same:
https://github.com/prisma/prisma/blob/edf305f55792a099bba340acffce37740cd030b3/packages/cli/src/utils/platform.ts#L91

However, you are right that consumer would need to remove the version number and/or add some extra string manipulation. I personally like the (1) more

Reply:
> @jasonkuhrt already established a pattern here and I thought to follow the same pattern

Then it makes sense to keep the same pattern. Based on that, should we then also declare `state` the same way?
```ts
const state = { userAgent: `prisma@<version>` }
```

I guess this is because we are limited to send this data only through params, instead of setting headers as you showed in the snippet right?

Reply:
> I guess this is because we are limited to send this data only through params, instead of setting headers as you showed in the snippet right?

Correct. We can't set the headers since we simply just open the address in the browser.

Reason why I was thinking to use name `client` is that it could become a pattern:
1. If console receives an external authentication request, the state is expected to include at least `client` and `redirectTo` properties